validator_client/src/key_cache.rs

```diff
@@ -0,0 +1,342 @@
+use account_utils::create_with_600_perms;
+use bls::{Keypair, PublicKey};
+use eth2_keystore::json_keystore::{
+    Aes128Ctr, ChecksumModule, Cipher, CipherModule, Crypto, EmptyMap, EmptyString, KdfModule,
+    Sha256Checksum,
+};
+use eth2_keystore::{
+    decrypt, default_kdf, encrypt, keypair_from_secret, Error as KeystoreError, Uuid, IV_SIZE,
+    SALT_SIZE,
+};
+use rand::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::OpenOptions;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+/// The file name for the serialized `KeyCache` struct.
+pub const CACHE_FILENAME: &str = "validator_key_cache.json";
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum State {
+    NotDecrypted,
+    DecryptedAndSaved,
+    DecryptedWithUnsavedUpdates,
+}
+
+fn not_decrypted() -> State {
+    State::NotDecrypted
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct KeyCache {
+    crypto: Option<Crypto>,
+    uuids: Vec<Uuid>,
+    #[serde(skip)]
+    pairs: HashMap<Uuid, Keypair>, //maps public keystore uuids to their corresponding Keypair
+    #[serde(skip)]
+    passwords: Vec<Vec<u8>>,
+    #[serde(skip)]
+    #[serde(default = "not_decrypted")]
+    state: State,
+}
+
+type SerializedKeyMap = HashMap<Uuid, Vec<u8>>;
```

Comment:
Changed it to ZeroizeHash since this map holds secret keys. Note that I needed to add Serialize and Deserialize derives to ZeroizeHash to be able to serialize the map into a byte vector that we can encrypt and to deserialize the byte vector after decryption.